### DISCUSSION TITLE
I get an error when using the custom reporter ERR_REQUIRE_ESM (TypeScript)

### DISCUSSION THREAD

**Discussion by procesor2017** [webdriverio/webdriverio#10647] [Q&A] [answered]
I am trying to create my own reporter according to the documentation in Typescript. The problem occurs in the import. When I try to import my reporter I get an error code: 'ERR_REQUIRE_ESM'.
I found this thread about the problem: https://github.com/webdriverio/webdriverio/discussions/10090 but nothing from it helped me. When I change the module to ESNext I just get another error: SyntaxError: Cannot use import statement outside a module.

My tsconfig.json:
```
{
    "compilerOptions": {
        "moduleResolution": "node",
        "module": "commonjs",
        "esModuleInterop": true,
        "forceConsistentCasingInFileNames": true,
        "baseUrl": ".",
        "paths": {
            "*": ["node_modules/*"]
          },
        "types": [
            "ts-node",
            "@wdio/globals/types",
            "expect-webdriverio",
            "@wdio/mocha-framework",
            
        ],
        "target": "es2022",
        
    },
    "include": ["src/**/*"],
    "exclude": ["node_modules"]

}
```
 package.json
 
```
{
{
  "devDependencies": {
    "@babel/cli": "^7.21.5",
    "@babel/core": "^7.22.1",
    "@babel/node": "^7.22.1",
    "@babel/preset-env": "^7.22.4",
    "@babel/register": "^7.21.0",
    "@testiny/cli": "^1.2.7",
    "@types/mssql": "^8.1.2",
    "@types/mysql": "^2.15.21",
    "@types/node": "^20.2.5",
    "@wdio/appium-service": "^8.10.0",
    "@wdio/cli": "^8.10.0",
    "@wdio/junit-reporter": "^8.11.0",
    "@wdio/local-runner": "^8.10.0",
    "@wdio/mocha-framework": "^8.10.0",
    "@wdio/reporter": "^8.11.0",
    "@wdio/sauce-service": "^8.11.2",
    "@wdio/spec-reporter": "^8.10.0",
    "axios": "^1.4.0",
    "chai": "^4.3.7",
    "junit-reporter": "^0.0.2",
    "msnodesqlv8": "^3.1.0",
    "mysql": "^2.18.1",
    "ts-node": "^10.9.1",
    "tslib": "^2.5.2",
    "typescript": "^5.0.4",
    "wdio-json-reporter": "^3.0.0"
  },
  "scripts": {
    "pretest": "export TEMPLATE_FILE=wdio.android.local.conf.ts && export NAME_OF_CONFIG=wdio.android.local.conf.ts && ts-node --transpile-only ./src/utils/helpers/pretest.ts",
    "pretest-sauce-android": "export TEMPLATE_FILE=sauce.android.config && export NAME_OF_CONFIG=wdio.android.sauce.conf.ts && ts-node --transpile-only ./src/utils/helpers/pretest.ts",
    "dashboard": "export LANGUAGE=de && wdio run ./conf/wdio.android.local.conf.ts --mochaOpts.grep Dashboard"
  }
}

``` 
customReporter.ts
```
import WDIOReporter from '@wdio/reporter';

export default class CustomReporter extends WDIOReporter {
    constructor(options) {
        /*
         * make reporter to write to the output stream by default
         */
        options = Object.assign(options, { stdout: true })
        super(options)
    }

    onTestPass(test) {
        this.write(`Congratulations! Your test "${test.title}" passed 👏`)
    }
}
``` 
wdio.conf:
```typescript
import { Options } from '@wdio/types';

import CustomReporter from './customReporter';

export const config: Options.Testrunner = {
    //
    // ====================
    // Runner Configuration
    // ====================
    // WebdriverIO supports running e2e tests as well as unit and component tests.
    runner: 'local',
    autoCompileOpts: {
        autoCompile: true,
        tsNodeOpts: {
            project: 'tsconfig.json',
            transpileOnly: true
        }
    },

    reporters: ['spec', 
                    ['junit', {
                        outputDir: './',
                        suiteNameFormat: /[^a-zA-Z0-9@>]+/,
                        outputFileFormat: function(options) {
                            return `results-2.xml`
                        },
                        testName: 'ahoj',
                        addFileAttribute: true,
                        packageName: ' ',
                }], CustomReporter
    ],
```

**Comment by christian-bromann** [comment]
> "module": "commonjs",

Please make sure your custom reporter is written as ESM module given that `@wdio/reporter` is still an ESM only package.

**Reply by procesor2017** [reply]
So if I understand correctly I have no choice but to have everything in the ESM module?
In that case then it should be in the documentation that if you want to use a custom reporter so you have to set:
package.json:
```
...
"type": "module"
...
``` 

tsconfig:
```
{
    "compilerOptions": {
        "module": "ES2022",
        "baseUrl": ".",
        "paths": {
            "*": ["node_modules/*", "*"]
          },
        "types": [
            "ts-node",
            "@wdio/globals/types",
            "expect-webdriverio",
            "@wdio/mocha-framework",
            
        ],
        "target": "es2022",
        "allowImportingTsExtensions": true,
        "noEmit": true   
        
    },
    "include": ["src/**/*"],
    "exclude": ["node_modules"]

}
``` 
and every import must be with ".ts"

Or is there a better option? I'm just starting out in typescript and I'm not sure how to ask google.

**Reply by christian-bromann** [reply]
> So if I understand correctly I have no choice but to have everything in the ESM module?

Are you writing a custom reporter for NPM or just for your test suite? I generally would recommend to write everything in ESM. I believe there is an open issue to make the reporter CJS compatible. Any contribution are welcome.

> and every import must be with ".ts"

not sure what you mean by that

**Reply by procesor2017** [reply]
> Are you writing a custom reporter for NPM or just for your test suite?

Just an ordinary reporter for my tests. I use Test manager tool which has only two options to import the result, either using API or CSV. So I have a reporter which creates csv and then uploads it to tsm.

I ask about ESM because I followed wdio documentation, which says that package.json must not contain modules and json should contain commonjs.

If this has changed in newer versions, i.e. it should be more like ES2022 (or another version) wouldn't it be desirable to include this in the documentation and also edit the documentation where there is a sample of how to write your own reporter? Because if you follow the documentation, it doesn't work on the latest version and if you are new to TS like me, you have a lot of searching to do.

 > not sure what you mean by that

After the changes I made:
package.json
```
{
  "devDependencies": {
    "@testiny/cli": "^1.2.7",
    "@types/mssql": "^8.1.2",
    "@types/mysql": "^2.15.21",
    "@types/node": "^20.2.5",
    "@wdio/appium-service": "^8.10.0",
    "@wdio/cli": "^8.10.0",
    "@wdio/junit-reporter": "^8.11.0",
    "@wdio/local-runner": "^8.10.0",
    "@wdio/mocha-framework": "^8.10.0",
    "@wdio/reporter": "^8.11.0",
    "@wdio/sauce-service": "^8.11.2",
    "@wdio/spec-reporter": "^8.10.0",
    "axios": "^1.4.0",
    "chai": "^4.3.7",
    "csv-writer": "^1.6.0",
    "esm": "^3.2.25",
    "junit-reporter": "^0.0.2",
    "msnodesqlv8": "^3.1.0",
    "mysql": "^2.18.1",
    "ts-node": "^10.9.1",
    "tsconfig-paths": "^4.2.0",
    "tslib": "^2.5.2",
    "typescript": "^5.0.4"
  },
  "scripts": {
    "pretest": "export TEMPLATE_FILE=wdio.android.local.conf.ts && export NAME_OF_CONFIG=wdio.android.local.conf.ts && ts-node ./src/utils/helpers/pretest.ts",
    "pretest-sauce-android": "export TEMPLATE_FILE=sauce.android.config && export NAME_OF_CONFIG=wdio.android.sauce.conf.ts && ts-node --transpile-only ./src/utils/helpers/pretest.ts",
    "dashboard": "export LANGUAGE=de && wdio run ./conf/wdio.android.local.conf.ts --mochaOpts.grep Dashboard"
  },
  "type": "module"
}
``` 

tsconfig.json

```
{
    "compilerOptions": {
        "module": "ES2022",
        "baseUrl": ".",
        "paths": {
            "@/*": ["node_modules/*", "*"]
          },
        "types": [
            "ts-node",
            "@wdio/globals/types",
            "expect-webdriverio",
            "@wdio/mocha-framework",
            
        ],
        "target": "es2022",
        "allowImportingTsExtensions": true,
        "noEmit": true   
        
    },
    "include": ["src", "*"],
    "exclude": ["node_modules"]

}
``` 

i must change import from:
```typescript
import inputSmsCodeComponent from "../../components/inputSmsCode.component";
import Database from '../../../../utils/helpers/database';

``` 
to:
```typescript
import inputSmsCodeComponent from "../../components/inputSmsCode.component.ts";
import Database from '../../../../utils/helpers/database.ts';
``` 

Without this change I couldn't get the imports to work.

**Reply by christian-bromann** [reply, marked as answer]
>  which says that package.json must not contain modules and json should contain commonjs.

Where does it say that exactly?

> If this has changed in newer versions, i.e. it should be more like ES2022 (or another version) wouldn't it be desirable to include this in the documentation and also edit the documentation where there is a sample of how to write your own reporter? 

Any contributions to the docs are appreciated.

> Without this change I couldn't get the imports to work.

Yes, in ESM you always need a file extension (`.js` would also work).

**Comment by ForesightImaging** [comment]
ugh, the update from 7 => 8 has been terrible. The typscript examples are broken, node-ts used to 'just work' for config and reporters. Now it's a mess. I'm not sure who this module transition benefits, but everything about using wdio got a lot harder.

**Reply by ForesightImaging** [reply]
I think the trouble here is that folks can't help update the doc because they can't make it work. 

I made a reporter to replace the default junit reporter because it cannot handle nested describes and we have had to migrate tools enough that it's not worth changing all the tests to accommodate the quirks of the reporter. I'm also running under electron, so grabbing images and video clips an integrating them into the report. 

All in typescript in wd7 just worked with ts-node out of the box. I have spent 2 days twiddling options, enabling top-level async re-working existing code with ridiculous promise all attempts to get imports that have always worked.

So a single working example of using ts-node to return a config that uses a custom reporter would likely help a lot of folks who obviously haven't had to deal with this before, but are forced to now.

**Reply by christian-bromann** [reply]
I would really have liked to accompany you through this process because we tried to make this as smooth as possible. Unfortunately given all the ways how people use WebdriverIO there is no guarantee that this is smooth for everyone which is why we rely on feedback and contributions from you to make the experience better.

To be clear: WebdriverIO v8 with TypeScript still **just** works and if you start a fresh WebdriverIO project via `npm init wdio@latest .` you can see this. I am starting WebdriverIO TypeScript projects almost every day and experience first when things break.

> I made a reporter to replace the default junit reporter because it cannot handle nested describes

Is this something you would like to contribute back to the project?

> enabling top-level async

I am not sure what needed to be re-enabled but top level await works since Node.js v14.8 and doesn't need any enabling. If you have a reproducible example, please share it.

**Reply by CedricLevasseur** [reply]
Hello
So i’m obliged to create a new project with wdio command ? I cannot create a Angular project with _ng new myproject_ ?

**Reply by christian-bromann** [reply]
Of course you can. Once you have Angular set up you can add a WebdriverIO setup with `npm init wdio .`

**Reply by ForesightImaging** [reply]
So to test if it 'works out of the box' I took your suggestion. 
npm init wdio@latest
Then added a single file for the custom reporter hello world class and imported it into the generated wdio.config and added the reporter class to the reporters entry ala.

reporter.ts
```
import WDIOReporter from '@wdio/reporter';

export default class CustomReporter extends WDIOReporter {
    constructor(options) {
        /*
         * make reporter to write to the output stream by default
         */
        options = Object.assign(options, { stdout: true })
        super(options)
    }

    onTestPass(test) {
        this.write(`Congratulations! Your test "${test.title}" passed 👏`)
    }
}
```

added to wdio.config.ts
```

   import CustomReporter from './reporter'
   
   reporters: ['spec','dot', CustomReporter],

```
The results:
```
C:\gitlab\wdioTestbed>npm run wdio

> wdio
> wdio run ./wdio.conf.ts

2023-08-18T21:26:57.654Z ERROR @wdio/config:ConfigParser: Failed loading configuration file: file:///C:/gitlab/wdioTestbed/wdio.conf.ts: Cannot find module 'C:\gitlab\wdioTestbed\reporter' imported from C:\gitlab\wdioTestbed\wdio.conf.ts
CustomError: Cannot find module 'C:\gitlab\wdioTestbed\reporter' imported from C:\gitlab\wdioTestbed\wdio.conf.ts
    at finalizeResolution (C:\gitlab\wdioTestbed\node_modules\ts-node\dist-raw\node-internal-modules-esm-resolve.js:366:11)
    at moduleResolve (C:\gitlab\wdioTestbed\node_modules\ts-node\dist-raw\node-internal-modules-esm-resolve.js:801:10)
    at Object.defaultResolve (C:\gitlab\wdioTestbed\node_modules\ts-node\dist-raw\node-internal-modules-esm-resolve.js:912:11)
    at C:\gitlab\wdioTestbed\node_modules\ts-node\src\esm.ts:218:35
    at entrypointFallback (C:\gitlab\wdioTestbed\node_modules\ts-node\src\esm.ts:168:34)
    at C:\gitlab\wdioTestbed\node_modules\ts-node\src\esm.ts:217:14
    at addShortCircuitFlag (C:\gitlab\wdioTestbed\node_modules\ts-node\src\esm.ts:409:21)    at resolve (C:\gitlab\wdioTestbed\node_modules\ts-node\src\esm.ts:197:12)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:841:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:40)
    at link (node:internal/modules/esm/module_job:75:36)
    ```

**Reply by christian-bromann** [reply]
In ESM you can't import without file extensions. To import your reporter correctly, use

```ts
import CustomReporter from './reporter.js'
```

**Reply by ForesightImaging** [reply]
well now, that's the kind of thing that belongs in the doc. I shall see if I can get this going and read a bit to understand why.

**Reply by ForesightImaging** [reply]
I've spent some time with this and re-worked our 'target' 'module' and 'moduleResolution' and updated to typescript 5.1. Unless I'm missing something else... when you use the .js trick. You get one level of import. If that 'js' referenced imports anything it has to use the same style imports.

**Reply by christian-bromann** [reply]
> when you use the .js trick

This is not a trick, it is what you actually want to do. In the CJS past it was ambiguous what you would import (yes, there was a clear order defined however it is not immediately clear to every developer). I recommend start importing `.js` files everywhere 😉

**Reply by ForesightImaging** [reply]
I was eventually able to get this to run, it however breaks every other place that js is used, like for webpack config and legacy code because it required the change to module in the package.json. (also any code that was common to the reporter and the product doesn't work anymore because of the .js import changes) Is there a way to get wdio to load a config file that imports WDIOReporter without setting     "type": "module"?

**Reply by ForesightImaging** [reply]
Dynamic imports would help if there was a way to async supply the export const config: Options.Testrunner.

**Reply by christian-bromann** [reply]
You can put all your WebdriverIO code in an `e2e` folder and put a `package.json` in there with this content:

```json
{
    "type": "module"
}
```
This way you can keep your application as CJS app and your WebdriverIO code ESM. Another option is to name all files `*.mts`.

**Reply by ForesightImaging** [reply]
I'll give that a go. I tried using an extension ala wdio.config.mts, but that file apparently has to be js or ts. I had hoped that the esm flag in ts-node might help as well, but nope.

**Reply by christian-bromann** [reply]
Let me test this but WebdriverIO should support `.mts` extensions. At least the latest version.

**Reply by ForesightImaging** [reply]
I'm currently on 8.13.13 because I believe that wdio-electron-service is broken on versions > than 8.13 due to change in the way that the chrome driver is resolved. I'm trying to get this sorted before I get that sorted.

**Reply by christian-bromann** [reply]
fyi: I raised a PR to make the reporter CJS compatible: https://github.com/webdriverio/webdriverio/pull/10998

**Reply by ForesightImaging** [reply]
thanks a lot. I think that is probably the best use of your time. Our project had not entered this corner of the javascript ecosystem before and since we got here via a change in webdriverio it's easily viewed as a problem with webdriverio rather than a ecosystem problem that I'm now more aware of. 

The folder level     "type": "module" has allowed me to move further. Though I've had to duplicate the shared source into that folder for the moment and manage the paths inside. I figure I'll just throw some more problems on the stack to pop later and see how deep I have to go before I can run with this new settings attempt.

Thanks again.

**Reply by ForesightImaging** [reply]
as a follow up, while I was able to get the wdio config to compile and my custom reporter to compile as well, those change then required changes to all my spec file imports. So I'm going to shelve this and urge wdio-electron-service to ensure > 8.13 support and await your recent change to get into the wild.

**Reply by christian-bromann** [reply]
> and await your recent change to get into the wild.

Already released https://github.com/webdriverio/webdriverio/releases/tag/v8.15.6
